sql: refactor casts and add volatility information

RaduBerinde opened the pull request:
#### sql: move code related to casts into separate file

Release note: None

#### sql: refactor casts and add volatility information

This change refactors the lists of valid casts into a single list and associated
lookup-up map. The new list defines casts between *families* of types, which was
effectively what was happening before but in a more roundabout way.

The casts now have volatility information, which will be used by the optimizer;
the list was obtained by auditing the usage of `ctx` in `PerformCast`.

Informs #48717.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49578)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-:-M8JPfjI4vkY_2KXxbeg:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @otan, and @RaduBerinde)

___
*[pkg/sql/sem/tree/casts.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8JOpv8Aepjgw9wjvzi:-M8JOpv8Aepjgw9wjvzj:bqadyw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L54)):*
> ```Go
> //    - session settings (e.g. bytes encoding format)
> //    - current timezone
> //    - current time (e.g. 'now'::string).
> ```

super nit: maybe `s/'now'/now()/g`?
___
*[pkg/sql/types/types.go, line 1168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8JO6WY-vGJnw_8iUUK:-M8JO6WZ-3kNz1sPur8u:b-ntgq7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/types/types.go#L1168)):*
> ```Go
> 	UnknownFamily:        "unknown",
> 	UuidFamily:           "uuid",
> 	EnumFamily:           "enum",
> ```

nit: seems like all families are in lexicographical order here except for `EnumFamily`, any reason for this?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49578)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/sql/sem/tree/casts.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LWWa72gkl_URgNXXg:-M8LWWa72gkl_URgNXXh:b-shv3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L56)):*
> ```Go
> //    - current time (e.g. 'now'::string).
> //
> var validCasts = []castInfo{
> ```

is this verified against PostgreSQL in any way?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/tree/casts.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LX8fgBZkYn7Vt0AzW:-M8LX8fhCFw8YnSWFwTJ:b27hmer) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L300)):*
> ```Go
> // types.T.
> func PerformCast(ctx *EvalContext, d Datum, t *types.T) (Datum, error) {
> 	switch t.Family() {
> ```

(not for this PR, perhaps if not already) is there a test that asserts all casts in the castsMap can be performed here?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/tree/casts.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LX8fgBZkYn7Vt0AzW:-M8LXJj8B6TUDOH_OdZI:b-6f399m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L300)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(not for this PR, perhaps if not already) is there a test that asserts all casts in the castsMap can be performed here?
</blockquote></details>

(in some ways, i wish `castsMap` contains the definition of the function as well, as it's easier to reason volatility if the definition is right next to it)


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/sem/tree/casts.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8JOpv8Aepjgw9wjvzi:-M8LYm1Z5qm8cx6gaX5s:bvgkeqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L54)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: maybe `s/'now'/now()/g`?
</blockquote></details>

now() is a function, it has its own volatility. this is about relative timestamp strings like 'now', 'tomorrow' (if you didn't know they were a thing, neither did I until a week or so ago)


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49578)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @otan, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/sem/tree/casts.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LWWa72gkl_URgNXXg:-M8LZGWl2vr-W6Wz1w3s:bdm22l4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L56)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

is this verified against PostgreSQL in any way?
</blockquote></details>

No.. maybe you can help me figure out where to get that information?
___
*[pkg/sql/sem/tree/casts.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LX8fgBZkYn7Vt0AzW:-M8LZOGN7uhMnmHUyWz4:bfqsq9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L300)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(in some ways, i wish `castsMap` contains the definition of the function as well, as it's easier to reason volatility if the definition is right next to it)
</blockquote></details>

I believe that we have tests for all casts, I definitely tripped some while working on thie PR.

I agree that we should be defining functions but I won't do that now. I will add a TODO.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/tree/casts.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LWWa72gkl_URgNXXg:-M8Lc1D55monBf8I4yze:bvka5fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L56)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

No.. maybe you can help me figure out where to get that information?
</blockquote></details>

`select c.castsource, c.casttarget, p.provolatile FROM pg_cast c JOIN pg_proc p ON (c.castfunc = p.oid);`
should be from_oid, to_oid, volatility.
you can the family "equivalence" code from https://github.com/cockroachdb/cockroach/blob/9de3eba85727276566d69ff0fcab8098069bfe78/pkg/sql/sem/builtins/all_builtins_test.go#L55


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/sem/tree/casts.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LWWa72gkl_URgNXXg:-M8LdrdwDupJEkDwd7G-:bm5cija) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L56)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

`select c.castsource, c.casttarget, p.provolatile FROM pg_cast c JOIN pg_proc p ON (c.castfunc = p.oid);`
should be from_oid, to_oid, volatility.
you can the family "equivalence" code from https://github.com/cockroachdb/cockroach/blob/9de3eba85727276566d69ff0fcab8098069bfe78/pkg/sql/sem/builtins/all_builtins_test.go#L55
</blockquote></details>

Oh, wow, you are awesome! I'll see if I can work that into a test.


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/sem/tree/casts.go:
```diff
@@ -0,0 +1,1012 @@
+	{from: types.JsonFamily, to: types.JsonFamily, volatility: VolatilityImmutable},
+
+	// Casts to EnumFamily.
+	{from: types.UnknownFamily, to: types.EnumFamily, volatility: VolatilityImmutable},
```
This is missing a bytes -> enum cast

RaduBerinde replied:
Will fix, I resolved conflicts with that change but I guess I added the opposite cast. Does seem like this is missing tests (I see you added a logictest in a subsequent PR but other casts have tests inside `sem/tree`).

rohany replied:
Yeah, i would like a test here, but it is a little tricky simulate the user defined type without the machinery creating it in the background. I'll spend some time making this easier to do.

rohany commented on pkg/sql/sem/tree/casts.go:
```diff
@@ -0,0 +1,1012 @@
+	{from: types.UnknownFamily, to: types.BytesFamily, volatility: VolatilityImmutable},
+	{from: types.StringFamily, to: types.BytesFamily, volatility: VolatilityImmutable},
+	{from: types.CollatedStringFamily, to: types.BytesFamily, volatility: VolatilityImmutable},
+	{from: types.EnumFamily, to: types.BytesFamily, volatility: VolatilityImmutable},
```
I don't think we're going to expose enum -> bytes for now (might revisit this in the future)

rohany reviewed (commented):
Swinging in because I landed something touched this a bit ago

RaduBerinde committed:
sql: refactor casts and add volatility information

This change refactors the lists of valid casts into a single list and associated
lookup-up map. The new list defines casts between *families* of types, which was
effectively what was happening before but in a more roundabout way.

The casts now have volatility information, which will be used by the optimizer;
the list was obtained by auditing the usage of `ctx` in `PerformCast`.

Informs #48717.

Release note: None

RaduBerinde committed:
sql: move code related to casts into separate file

Release note: None

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49578)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @otan, @rohany, and @yuzefovich)

___
*[pkg/sql/sem/tree/casts.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8LWWa72gkl_URgNXXg:-M8NFO8a5ciePnpTOJmm:bhc1rqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L56)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Oh, wow, you are awesome! I'll see if I can work that into a test.
</blockquote></details>

Done. Only one exception! (though a lot of our casts don't have a pg equivalent).
___
*[pkg/sql/sem/tree/casts.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8JDKCOn2G-8bRkp28z-r2-184:-M8NFUn17d4dfA8No1bm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L184)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I don't think we're going to expose enum -> bytes for now (might revisit this in the future)
</blockquote></details>

Done.
___
*[pkg/sql/types/types.go, line 1168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8JO6WY-vGJnw_8iUUK:-M8NFdmL4EL0Yqft-mTo:b265m2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/types/types.go#L1168)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: seems like all families are in lexicographical order here except for `EnumFamily`, any reason for this?
</blockquote></details>

Fixed. I think I copied them one by one from the T.Name() switch.


<!-- Sent from Reviewable.io -->

otan reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49578)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @otan, @rohany, and @yuzefovich)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-:-M8Ne7hY3Acf4Dl4eK0y:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @rohany, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49578)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @RaduBerinde, @rohany, and @yuzefovich)

___
*[pkg/sql/sem/tree/casts.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49578#-M8JOpv8Aepjgw9wjvzi:-M8NpBBfCk4PvpKAwtU7:b828jsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec4114c4b24059af5560433c995b9511c41afa5/pkg/sql/sem/tree/casts.go#L54)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

now() is a function, it has its own volatility. this is about relative timestamp strings like 'now', 'tomorrow' (if you didn't know they were a thing, neither did I until a week or so ago)
</blockquote></details>

Yeah, I did not know that.


<!-- Sent from Reviewable.io -->
